File: frontend/lib/models/vehicle.dart
import 'doc_ref.dart';
import 'enums.dart';
import 'gated_entity.dart';

/// An auto-rickshaw in inventory. Gated: status `active` shows as "Verified".
/// "Assigned" in the UI means it is tied to a customer (via a sale or rental).
class Vehicle with GatedEntity {
  Vehicle({
    required this.id,
    required this.regNo,
    required this.type,
    this.branch,
    this.purchaseDate,
    this.chassisNo,
    this.model,
    this.fuelType,
    this.buyingExpenses,
    this.showroom,
    this.rcPermit = false,
    this.insuranceDate,
    this.fcDate,
    this.permitDate,
    Map<VehicleDocType, String>? documents,
    this.prevOwnerName,
    this.prevOwnerMobile,
    this.prevOwnerAddress,
    this.prevOwnerIdProof,
    this.prevOwnerPhoto,
    this.assignedToCustomerId,
    this.inventoryStatus = InventoryStatus.available,
    this.saleStatus = SaleStatus.notSold,
    this.nextServiceDueDate,
    required this.createdBy,
    required this.createdAt,
    this.status = EntityStatus.active,
    this.confirmedBy,
    this.confirmedAt,
    this.rejectionReason,
    List<DocRef>? uploadedDocs,
  })  : documents = documents ?? {},
        uploadedDocs = uploadedDocs ?? [];

  @override
  final String id;

  /// Registration number. Captured for second-hand vehicles; a first-hand
  /// (new) vehicle is unregistered, so this may be empty until registration.
  String regNo;
  VehicleType type;
  Branch? branch;
  DateTime? purchaseDate;

  // Common details (both hand types).
  String? chassisNo;
  String? model;
  FuelType? fuelType;
  num? buyingExpenses;

  // First-hand only: source showroom + RC permit flag.
  Showroom? showroom;
  bool rcPermit;

  // Second-hand only: existing paperwork dates + uploaded documents.
  DateTime? insuranceDate;
  DateTime? fcDate;
  DateTime? permitDate;

  /// Uploaded documents: doc type → file name (path/URL in Phase 7 backend).
  Map<VehicleDocType, String> documents;

  // Second-hand only: previous-owner (buyer) details. ID proof / photo hold the
  // uploaded file name (path/URL in Phase 7); null = not attached.
  String? prevOwnerName;
  String? prevOwnerMobile;
  String? prevOwnerAddress;
  String? prevOwnerIdProof;
  String? prevOwnerPhoto;

  /// Customer this vehicle is currently assigned to (sale or rental), if any.
  String? assignedToCustomerId;
  InventoryStatus inventoryStatus;
  SaleStatus saleStatus;

  /// Documents already stored on the backend (for download / replace / delete).
  List<DocRef> uploadedDocs;

  /// Denormalised next-service date (most recent service + interval). Drives
  /// the service-due banner in the rental module.
  DateTime? nextServiceDueDate;

  @override
  final String createdBy;
  @override
  final DateTime createdAt;
  @override
  EntityStatus status;
  @override
  String? confirmedBy;
  @override
  DateTime? confirmedAt;
  @override
  String? rejectionReason;

  bool get isAssigned => assignedToCustomerId != null;

  /// Display id, e.g. `VEH-001` from `v_001` or the raw id.
  String get displayId {
    final digits = RegExp(r'\d+').firstMatch(id)?.group(0);
    return digits != null ? 'VEH-${digits.padLeft(3, '0')}' : id.toUpperCase();
  }

  /// Human-readable vehicle label. Uses reg number when available (second-hand);
  /// falls back to model or type+ID for first-hand (unregistered) vehicles.
  String get displayLabel =>
      regNo.isNotEmpty ? regNo : (model ?? '${type.label} #$displayId');
}

File: backend/app/schemas/vehicle.py
"""Pydantic request/response schemas for the vehicles module."""
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict

from app.models.enums import (
    Branch,
    EntityStatus,
    FuelType,
    InventoryStatus,
    SaleStatus,
    Showroom,
    VehicleDocType,
    VehicleHand,
)


class VehicleBase(BaseModel):
    # protected_namespaces=() so the `model` field doesn't clash with pydantic
    model_config = ConfigDict(protected_namespaces=())

    branch: Branch
    owned_hand: VehicleHand
    chassis_no: str
    model: str
    fuel_type: FuelType | None = None
    purchase_date: date | None = None
    buying_expenses: Decimal | None = None
    # first-hand
    showroom: Showroom | None = None
    # RC / Permit / Insurance — independent flags (both hand types)
    rc: bool = False
    permit: bool = False
    insurance: bool = False
    # second-hand
    reg_no: str | None = None
    insurance_date: date | None = None
    fc_date: date | None = None
    permit_date: date | None = None
    prev_owner_name: str | None = None
    prev_owner_mobile: str | None = None
    prev_owner_address: str | None = None
    remarks: str | None = None
    financer_id: int | None = None


class VehicleCreate(VehicleBase):
    module_code: str = "auto_sale"
    status: EntityStatus | None = None  # optional override; else derived from role


class VehicleUpdate(BaseModel):
    """All fields optional — only provided keys are changed (PATCH)."""
    model_config = ConfigDict(protected_namespaces=())

    branch: Branch | None = None
    owned_hand: VehicleHand | None = None
    chassis_no: str | None = None
    model: str | None = None
    fuel_type: FuelType | None = None
    purchase_date: date | None = None
    buying_expenses: Decimal | None = None
    showroom: Showroom | None = None
    rc: bool | None = None
    permit: bool | None = None
    insurance: bool | None = None
    reg_no: str | None = None
    insurance_date: date | None = None
    fc_date: date | None = None
    permit_date: date | None = None
    prev_owner_name: str | None = None
    prev_owner_mobile: str | None = None
    prev_owner_address: str | None = None
    sale_status: SaleStatus | None = None
    status: EntityStatus | None = None
    remarks: str | None = None
    financer_id: int | None = None


class DocumentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    doc_type: VehicleDocType
    file_name: str
    mime_type: str
    size_bytes: int | None = None
    uploaded_at: datetime


class VehicleOut(VehicleBase):
    model_config = ConfigDict(from_attributes=True, protected_namespaces=())

    id: int
    module_id: int
    inventory_status: InventoryStatus
    sale_status: SaleStatus
    is_seized: bool = False
    next_service_due_date: date | None = None
    assigned_to_customer_id: int | None = None
    status: EntityStatus
    created_by: int
    created_at: datetime
    confirmed_by: int | None = None
    confirmed_at: datetime | None = None
    rejection_reason: str | None = None
    documents: list[DocumentOut] = []
